/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

syntax = "proto3";
package teaclave_authentication_service_proto;

import "teaclave_common.proto";

message UserRegisterRequest {
  string id = 1;
  string password = 2;
  string role = 3;
  string attribute = 4;
}

message UserRegisterResponse { }

message UserUpdateRequest {
    string id = 1;
    string password = 2;
    string role = 3;
    string attribute = 4;
}

message UserUpdateResponse { }

message UserLoginRequest {
  string id = 1;
  string password = 2;
}

message UserLoginResponse {
  string token = 1;
}

message UserAuthenticateRequest {
  teaclave_common_proto.UserCredential credential = 1;
}

message UserAuthClaims {
  string sub = 1;
  string role = 2;
  string iss = 3;
  uint64 exp = 4;
}

message UserAuthenticateResponse {
  UserAuthClaims claims = 1;
}

service TeaclaveAuthenticationApi {
  rpc UserRegister(UserRegisterRequest) returns (UserRegisterResponse);
  rpc UserUpdate(UserUpdateRequest) returns (UserUpdateResponse);
  rpc UserLogin (UserLoginRequest) returns (UserLoginResponse);
}

service TeaclaveAuthenticationInternal {
  rpc UserAuthenticate (UserAuthenticateRequest) returns (UserAuthenticateResponse);
}